### PR TITLE
Update polkadot inflation to take into account auctions

### LINKED ISSUE #3011: Track the number of auctioned slots which are currently active
For gilts, we want to know the number of slots which were won by auctions in the current block.

Some kind of tracking vector needs to be used for each of the different lease periods.

### PR DESCRIPTION
**Pull request by kianenigma**
fixes https://github.com/paritytech/polkadot/issues/5195
We should probably also close #3011 as we move to Polkadot, as the current approach is kind of a guess, without taking into account the number of slots that where actually auctioned.

Not exactly sure if this makes sense, as it was added mixed with the guilts PR (https://github.com/paritytech/polkadot/pull/2587), which polkadot does not have. 

Also related: https://github.com/paritytech/substrate/issues/8327

### CODE REVIEW

**Comment by gavofyork** [comment]
Looks ok - we might eventually want the economics of Polkadot to diverge from Kusama and thus need greater parameterisation with the `common` `era_payout` function. But it's probably fine for now.

File: runtime/common/src/impls.rs [hunk 1]
```diff
@@ -57,6 +59,45 @@ where
...
+	auctioned_slots: u64,
+) -> (Balance, Balance) {
+	use pallet_staking_reward_fn::compute_inflation;
+	use sp_runtime::traits::Saturating;
+
+	let min_annual_inflation = Perquintill::from_rational(25u64, 1000u64);
+	let delta_annual_inflation = max_annual_inflation.saturating_sub(min_annual_inflation);
+
+	// 30% reserved for up to 60 slots.
+	let auction_proportion = Perquintill::from_rational(auctioned_slots.min(60), 200u64);
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply runtime/common/src/impls.rs:76]: No - it just sets an effective upper limit to how much of the inflation can be accounted for by the slot auctions.

**Comment by bkchr** [comment]
@kianenigma you want to move forward with this?

**Comment by kianenigma** [comment]
> @kianenigma you want to move forward with this?

Yes, I made one attempt with my current commits that more or less works, but it needs a shitty data seeding. After chatting a bit with @KiChjang about it, we concluded that it can probably be done better. But I can't yet find the correct storage items that I need onchain, if they exist. 

In short, we want a list of "active para ids that are there through an auction". 

@shawntabrizi is probably a better person to guide me here, but I will nonetheless continue. 

The staking rate in Polkadot and Kusama is not off because of this.

**Comment by shawntabrizi** [comment]
I don't see this as the right immediate answer. The first thing I would do is create a new storage in the paras pallet to keep track of all the parachains which are active in a lease period.

From that, we will know how many parachains we have each lease period has. That seems like a better number to start with than what you have done here, which is not clear to me is even right, since this list never has items removed when they are offboarded.

**Comment by kianenigma** [comment]
> That seems like a better number to start with than what you have done here, which is not clear to me is even right, since this list never has items removed when they are offboarded.

The new list is never cleared, and it is supposed to represent the para-id of all parachains that have been offboarded. Then, we take the intersection of this with the current list of active para-ids, and that's the right number.

I actually think this is "correct", but perhaps what you are suggesting is easier.

**Comment by shawntabrizi** [comment]
So if a parachain is auctioned, then offboards, then becomes a common good chain...

And does this handle the same parachain winning 2 auctions? AFAIK it duplicates data then?

Not convinced this is really right

**Comment by kianenigma** [comment]
> So if a parachain is auctioned, then offboards, then becomes a common good chain...

Good point, this is not handled. A common good para is no longer locking any DOT and should not be tracked. 

> And does this handle the same parachain winning 2 auctions? AFAIK it duplicates data then?

The data in the new storage item is duplicate, but it will easily be deduplicated by making sure in the final filter, we do `active_para_ids.filter(|i| auction_winners_with_dupe.contains(i))` so even if a para-id has won multiple auctions, it is taken into account. 

---

I will revise my approach nonetheless.

File: runtime/kusama/src/lib.rs [hunk 2]
```diff
@@ -504,59 +504,32 @@ impl pallet_bags_list::Config<VoterBagsListInstance> for Runtime {
...
 		_total_issuance: Balance,
 		era_duration_millis: u64,
 	) -> (Balance, Balance) {
-		// TODO: #3011 Update with proper auctioned slots tracking.
-		// This should be fine for the first year of parachains.
-		let auctioned_slots: u64 = auctions::Pallet::<Runtime>::auction_counter().into();
+		let auctioned_slots = {
+			// all para-ids that have been leased out.
+			let auctioned_winner_para_ids = Auctions::auctioned_winner_para_ids();
+			// and all para-ids that are not active.
```
Comment (KiChjang) [inline runtime/kusama/src/lib.rs:517]:
```suggestion
			// and all para-ids that are currently active.
```

File: runtime/polkadot/src/lib.rs [hunk 3]
```diff
@@ -536,6 +536,42 @@ type SlashCancelOrigin = EitherOfDiverse<
...
+impl pallet_staking::EraPayout<Balance> for EraPayout {
+	fn era_payout(
+		total_staked: Balance,
+		_total_issuance: Balance,
+		era_duration_millis: u64,
+	) -> (Balance, Balance) {
+		let auctioned_slots = {
+			// all para-ids that have been leased out.
+			let auctioned_winner_para_ids = Auctions::auctioned_winner_para_ids();
+			// and all para-ids that are not active.
```
Comment (KiChjang) [inline runtime/polkadot/src/lib.rs:549]:
```suggestion
			// and all para-ids that are currently active.
```

**Comment by KiChjang** [comment]
So as I mentioned on Element, it would seem to me that the quick and easy way to do this is simply by doing `active_para_ids.filter(|i| i >= PUBLIC_INDEX_START)`, as we know for certain that `ParaId`s that are >= 2000 are done through auctions, hence there's no need to look into the `auction_winners_para_ids` storage item.

**Review by shawntabrizi — APPROVED** [review]

**Review by KiChjang — APPROVED** [review]

File: runtime/polkadot/src/lib.rs [hunk 4]
```diff
@@ -536,6 +537,35 @@ type SlashCancelOrigin = EitherOfDiverse<
...
+			.filter(|i| *i >= LOWEST_PUBLIC_ID)
+			.count() as u64;
+
+		const MAX_ANNUAL_INFLATION: Perquintill = Perquintill::from_percent(10);
+		const MILLISECONDS_PER_YEAR: u64 = 1000 * 3600 * 24 * 36525 / 100;
+
+		runtime_common::impls::era_payout(
+			total_staked,
+			// Polkadot has no notion of gilts, the entire issuance is non-guilt.
+			Balances::total_issuance(),
```
Comment (kianenigma) [inline runtime/polkadot/src/lib.rs:561]: Should I not use the `_total_issuance` here?

Comment (kianenigma) [reply runtime/polkadot/src/lib.rs:561]: Yes I should. https://github.com/paritytech/substrate/blob/fca01db6337fd471008875a2bfd9237b12aa9e66/frame/staking/src/pallet/impls.rs#L400

File: runtime/kusama/src/lib.rs [hunk 5]
```diff
@@ -504,59 +505,25 @@ impl pallet_bags_list::Config<VoterBagsListInstance> for Runtime {
...
 		era_duration_millis: u64,
 	) -> (Balance, Balance) {
-		// TODO: #3011 Update with proper auctioned slots tracking.
-		// This should be fine for the first year of parachains.
-		let auctioned_slots: u64 = auctions::Pallet::<Runtime>::auction_counter().into();
+		// all para-ids that are currently active.
+		let auctioned_slots = Paras::parachains()
+			.into_iter()
+			// all active para-ids that do not belong to a system or common good chain is the number
+			// of parachains that we should take into account for inflation.
```
Comment (gavofyork) [inline runtime/kusama/src/lib.rs:519]: is this true? what about parathreads which did not secure an auction slot?

File: runtime/polkadot/src/lib.rs [hunk 6]
```diff
@@ -536,6 +537,35 @@ type SlashCancelOrigin = EitherOfDiverse<
...
+impl pallet_staking::EraPayout<Balance> for EraPayout {
+	fn era_payout(
+		total_staked: Balance,
+		total_issuance: Balance,
+		era_duration_millis: u64,
+	) -> (Balance, Balance) {
+		// all para-ids that are not active.
+		let auctioned_slots = Paras::parachains()
+			.into_iter()
+			// all active para-ids that do not belong to a system or common good chain is the number
```
Comment (gavofyork) [inline runtime/polkadot/src/lib.rs:550]: ditto.

File: runtime/kusama/src/lib.rs (hunk 5 above)
Comment (shawntabrizi) [reply runtime/kusama/src/lib.rs:519]: the `::parachains()` api only returns the list of paras which have upgraded to a parachain. It does not include parathreads.

**Review by bkchr — APPROVED** [review]

**Comment by kianenigma** [comment]
merging speculatively to meet deadlines, def. need to wait for audit before enactment.

**Comment by kianenigma** [comment]
bot merge

**Comment by kianenigma** [comment]
Ran some experiments on Polkadot, numbers add up. As expected, rewards will be more: 

```
[runtime/polkadot/src/lib.rs:2442] DOT::from(total_issuance) = 1254311394,594 DOT (12,543,113,945,943,955,614)
[runtime/polkadot/src/lib.rs:2443] DOT::from(total_staked) = 614859453,927 DOT (6,148,594,539,274,007,157)

// in this PR
[runtime/polkadot/src/lib.rs:2446] DOT::from(validators) = 301672,607 DOT (3,016,726,077,584,039)
[runtime/polkadot/src/lib.rs:2447] DOT::from(remainder) = 41739,136 DOT (417,391,361,414,989)

// before this PR
[runtime/polkadot/src/lib.rs:2450] DOT::from(validators) = 254192,206 DOT (2,541,922,068,857,701)
[runtime/polkadot/src/lib.rs:2451] DOT::from(remainder) = 89219,438 DOT (892,194,382,832,565)
```

**Pull request merged**